pkg/expr/expr.go

```diff
@@ -0,0 +1,365 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expr
+
+import (
+	"bytes"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"strconv"
+	"strings"
+
+	"github.com/golang/glog"
+
+	config "istio.io/api/mixer/v1/config/descriptor"
+	"istio.io/mixer/pkg/attribute"
+)
+
+// This private variable is an extract from go/token
+// it maps expression related token to its names
+var tMap = map[token.Token]string{
+	token.ILLEGAL: "ILLEGAL",
+
+	token.EOF:     "EOF",
+	token.COMMENT: "COMMENT",
+
+	token.IDENT:  "IDENT",
+	token.INT:    "INT",
+	token.FLOAT:  "FLOAT",
+	token.IMAG:   "IMAG",
+	token.CHAR:   "CHAR",
+	token.STRING: "STRING",
+
+	token.ADD: "ADD",
+	token.SUB: "SUB",
+	token.MUL: "MUL",
+	token.QUO: "QUO",
+	token.REM: "REM",
+
+	token.AND:  "AND",
+	token.OR:   "OR",
+	token.XOR:  "XOR",
+	token.LAND: "LAND",
+	token.LOR:  "LOR",
+
+	token.EQL: "EQ",
+	token.LSS: "LT",
+	token.GTR: "GT",
+	token.NOT: "NOT",
+
+	token.NEQ: "NEQ",
+	token.LEQ: "LEQ",
+	token.GEQ: "GEQ",
+
+	token.LBRACK: "INDEX",
+}
+
+var typeMap = map[token.Token]config.ValueType{
+	token.INT:    config.INT64,
+	token.FLOAT:  config.DOUBLE,
+	token.CHAR:   config.STRING,
+	token.STRING: config.STRING,
+}
+
+// Expression is a simplified expression AST
+type Expression struct {
+	// Oneof the following
+	Const *Constant
+	Var   *Variable
+	Fn    *Function
+}
+
+// Eval evaluates the expression given an attribute bag and a function map.
+func (e *Expression) Eval(attrs attribute.Bag, fMap map[string]Func) (interface{}, error) {
+	if e.Const != nil {
+		return e.Const.TypedValue, nil
+	}
+	if e.Var != nil {
+		v, ok := attribute.Value(attrs, e.Var.Name)
+		if !ok {
+			return nil, fmt.Errorf("unresolved attribute %s", e.Var.Name)
+		}
+		return v, nil
+	}
+	// may panic
+	return e.Fn.Eval(attrs, fMap)
+}
+
+// String produces postfix version with all operators converted to function names
+func (e *Expression) String() string {
+	if e.Const != nil {
+		return e.Const.String()
+	}
+	if e.Var != nil {
+		return e.Var.String()
+	}
+	if e.Fn != nil {
+		return e.Fn.String()
+	}
+	return "<nil>"
+}
+
+// newConstant creates a new constant of given type.
+// It also stores a typed form of the constant.
+func newConstant(v string, vType config.ValueType) (*Constant, error) {
+	var typedVal interface{}
+	var err error
+	switch vType {
+	case config.INT64:
+		if typedVal, err = strconv.ParseInt(v, 10, 64); err != nil {
+			return nil, err
+		}
+	case config.DOUBLE:
+		if typedVal, err = strconv.ParseFloat(v, 64); err != nil {
+			return nil, err
+		}
+	default: // string
+		if typedVal, err = strconv.Unquote(v); err != nil {
+			return nil, err
+		}
+	}
+	return &Constant{Value: v, Kind: vType, TypedValue: typedVal}, nil
+}
+
+// Constant models a typed constant.
+type Constant struct {
```

Comment:
The field's names feel off here; I think it'd make more sense as:

    struct Constant struct {
        Str String
        Value interface{}
        Type config.ValueType
    }